db/src/main/java/org/odk/collect/db/sqlite/SynchronizedDatabaseConnection.kt

```diff
@@ -0,0 +1,46 @@
+package org.odk.collect.db.sqlite
+
+import android.content.Context
+import android.database.sqlite.SQLiteDatabase
+import androidx.core.database.sqlite.transaction
+
+class SynchronizedDatabaseConnection(
+    context: Context,
+    path: String,
+    name: String,
+    migrator: DatabaseMigrator,
+    databaseVersion: Int
+) {
+    private val databaseConnection = DatabaseConnection(
+        context,
+        path,
+        name,
+        migrator,
+        databaseVersion
+    )
+
+    fun <T> withConnection(block: DatabaseConnection.() -> T): T {
+        return databaseConnection.withSynchronizedConnection(block)
+    }
+
+    fun <T> transaction(
+        body: SQLiteDatabase.() -> T
+    ) {
+        return withConnection {
+            writeableDatabase.transaction {
+                body()
+            }
+        }
+    }
+
+    /**
+     * Runs a transaction and then calls [DatabaseConnection.reset]. Useful for transactions
+     * that will mutate the DB schema.
+     */
+    fun <T> resetTransaction(
+        body: SQLiteDatabase.() -> T
+    ) {
+        transaction(body)
+        databaseConnection.reset()
```

Comment:
I'd been unable to write a Robolectric test that would drive out adding this `reset()` with the existing implementation. I'm guessing that it may have been possible to write one in an instrumentation test. 

However, now if you remove `reset()`, 5 a few tests fail in the existing `DatabaseEntitiesRepository`. This is because the change to looking up columns to determine which to add (using `ALTER TABLE`) is much more sensitive to the problem than the previous implementation that used a `try-catch` to just "fire and forget" add columns. I'm leaning towards not bothering with a new instrumentation test as it feels unlikely that anyone would go back to the old hacky implementation, but happy to debate that!